centrifuge/src/main/java/io/github/centrifugal/centrifuge/Client.java

```diff
@@ -395,34 +395,67 @@ private void handleConnectionOpen() {
         if (this.getState() != ClientState.CONNECTING) {
             return;
         }
-        if (this.refreshRequired || (this.token.equals("") && this.opts.getTokenGetter() != null)) {
-            ConnectionTokenEvent connectionTokenEvent = new ConnectionTokenEvent();
-            if (this.opts.getTokenGetter() == null) {
-                this.listener.onError(Client.this, new ErrorEvent(new ConfigurationError(new Exception("tokenGetter function should be provided in Client options to handle token refresh, see Options.setTokenGetter"))));
-                this.processDisconnect(DISCONNECTED_UNAUTHORIZED, "unauthorized", false);
-                return;
-            }
-            this.opts.getTokenGetter().getConnectionToken(connectionTokenEvent, (err, token) -> this.executor.submit(() -> {
-                if (Client.this.getState() != ClientState.CONNECTING) {
+        if (this.refreshRequired) {
+            if (this.data == null && this.opts.getDataGetter() != null) {
```

Comment:
Token update may be lost in this case or maybe I am not following the code correctly?

Reply:
Sorry about the long delayed response. I think that it can't lost, because the logic is identical if the authorization token was set through the `setToken()` method. Only the “place of transfer” of the token is different. 

Извините за долгий ответ. Я не думаю, что токен может быть потерян, потому что логика идентична, если бы мы записали токен через метод `setToken()`. Отличается только "место хранения" токена.

Reply:
I mean that if `getDataGetter` is not null, then `getConnectionToken` won't be called.

Reply:
And another concern - data callback should be called every time if set during connection opening, see how it was implemented here as example - https://github.com/centrifugal/centrifuge-dart/pull/88/files

Reply:
> I mean that if getDataGetter is not null, then getConnectionToken won't be called.
that's right, then getConnectionData will be called and inside it you will need to update the token (exactly the same as in getConnectionToken)

>And another concern - data callback should be called every time if set during connection opening, see how it was implemented here as an example
I need some time to think. I still don’t understand where I should fix this